client/state/data-layer/http-data.js

```diff
@@ -184,8 +199,70 @@ export const requestHttpData = ( requestId, fetchAction, { fromApi, freshness =
 	return data;
 };
 
+/**
+ * Blocks execution until requested data has been fulfilled
+ *
+ *  - May return without data if data hasn't been fulfilled or failed
+ *  - Use for SSR contexts or when we _want_ to block rendering or execution
+ *    until the requested data has been fulfilled, e.g. when URL routing
+ *    depends on some data property
+ *  - _DO NOT USE_ when normal synchronous/data interactions suffice such
+ *    as is the case in 99.999% of React component contexts
+ *
+ * @example:
+ * withData( {
+ *     geo: () => requestGeoLocation(),
+ *     splines: () => requestSplines( siteId ),
+ * } ).then( ( { geo, splines } ) => {
+ *     return ( geo.state === 'success' || splines.state === 'success' )
+ *         ? res.send( renderToStaticMarkup( <LocalSplines geo={ geo.data } splines={ splines.data } /> ) )
+ *         : res.send( renderToStaticMarkup( <UnvailableData /> ) );
+ * }
+ *
+ * @param {object} query key/value pairs of data name and request
+ * @param {int} timeout how many ms to wait until giving up on requests
+ * @return {Promise<object>} fulfilled data of request (or partial if could not fulfill)
+ */
+export const withData = ( query, { timeout = 5000 } = {} ) =>
+	new Promise( ( resolve, reject ) => {
+		let unsubscribe = () => {};
+		let timer = null;
+		const names = Object.keys( query );
+
+		const getValues = () =>
+			names.reduce(
+				( [ values, allGood ], name ) => {
+					const value = query[ name ]();
+
+					return [ { ...values, [ name ]: value }, allGood && value.state === 'success' ];
+				},
+				[ {}, true ]
+			);
+
+		const listener = () => {
+			const [ values, allGood ] = getValues();
```

Comment:
In general case, it's impossible to infer the `allGood` value just from the stream of state updates. Resolution of the promise needs to be bound to end of the actual request. See p4TIVU-96L-p2 and the rich discussion there for detailed examples.

Said that, this approach works now, but will break down in the future once we add features like sending creation requests to create new items or "fetch policies" where some requests want just the cached data and some want to refresh from the network immediately.

Reply:
> it's impossible to infer the allGood value just from the stream of state updates.

technically you are right and I thought about this but wanted to keep the PR small.

> Resolution of the promise needs to be bound to end of the actual request.

here I disagree because if we do that then there's no use in the timeout and I felt that it's important enough to have a timeout because we don't want to make the application crash just because some random API isn't responding. we can set an infinite timeout if we want to truly block forever.

in the meantime, my next steps, which I think I would prefer not to do in this patch but maybe I will, is check for `allFailed` which is the only other non-timeout related abort case: either they all succeeded and we know to proceed or they all failed and we know we cannot proceed by waiting any longer